server/game/core/ongoingEffect/OngoingEffect.ts

```diff
@@ -29,7 +34,24 @@ const EnumHelpers = require('../utils/EnumHelpers');
  *                        and the numerical value of the effect, if any.
  */
 class OngoingEffect {
-    constructor(game, source, properties, effectImpl) {
+    public game: Game;
+    public source: Card;
+    // TODO: Can we make GameObject more specific? Can we add generics to the class for AbilityContext?
+    public matchTarget: GameObject | ((target: GameObject, context: AbilityContext) => boolean);
+    public duration?: Duration;
+    public until: WhenType;
+    public condition: (context?: AbilityContext) => boolean;
+    public sourceLocationFilter: LocationFilter;
+    public canChangeZoneOnce: boolean;
+    public canChangeZoneNTimes: number;
+    public impl: OngoingEffectImpl<any>;
+    // ISSUE: refreshContext sets ability to IOngoingEffectProps, but the listed type for context is PlayerOrCardAbility. Why is there a mismatch? Are we just overriding it in the context of OngoingEffects and everywhere else it acts as PlayerOrCardAbility?
```

Comment:
That was an oversight most likely, constant abilities are just an interface (a set of properties really) and don't actually derive from PlayerOrCardAbility. So it almost certainly is supposed to always be IOngoingEffectProps

Reply:
Yeah that's a little bit of a hack but I believe it's one propagated forward from jigoku. The only place we use it currently I believe is in `GainAbility.ts` which is the effect type used when a card gains an ability from the effect of another card, e.g. when an upgrade says "attached unit gains: `When played: ...`"

In there we have to refer back to the source ability that caused the gain to happen, and since context.ability (iirc) isn't naturally set for a constant ability / ongoing effect, we explicitly add it on a refresh. I _think_ that's what's going on from my memory, if that doesn't line up feel free to ping me about it, and if you can see a better pattern for accomplishing that then please definitely feel free to implement it

Reply:
Ok so I've rewritten this comment ~~4~~5 times due to different rabbit holes, but looking at jigoku it definitely wasn't propagated forward. this field was originally `this.ability = properties.ability` rather than the properties object itself, and refreshContext explicitly had `if(this.ability) { this.context.ability = this.ability }`. 

That said, the system definitely relies on its current design, I see the process of building the ongoing effect essentially makes the properties object ability-like. I'm just gonna leave type cast with a note and put this under "needs a look at another time".